src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -63,6 +67,30 @@ def _build_general_name(backend, gn):
     if gn.type == backend._lib.GEN_DNS:
         data = backend._ffi.buffer(gn.d.dNSName.data, gn.d.dNSName.length)[:]
         return x509.DNSName(idna.decode(data))
+    elif gn.type == backend._lib.GEN_URI:
+        data = backend._ffi.buffer(
+            gn.d.uniformResourceIdentifier.data,
+            gn.d.uniformResourceIdentifier.length
+        )[:].decode("ascii")
+        parsed = urllib_parse.urlparse(data)
+        hostname = idna.decode(parsed.hostname)
+        if parsed.port:
+            netloc = hostname + u":" + six.text_type(parsed.port)
+        else:
+            netloc = hostname
+
+        # Note that building a URL in this fashion means it should be
+        # semantically indistinguishable from the original but is not
+        # guaranteed to be exactly the same.
+        uri = urllib_parse.urlunparse((
+            parsed.scheme,
+            netloc,
+            parsed.path,
+            parsed.params,
+            parsed.query,
+            parsed.fragment
+        ))
```

Comment:
Is that a problem?

On Thu, Apr 23, 2015 at 8:30 AM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/backends/openssl/x509.py
> <https://github.com/pyca/cryptography/pull/1862#discussion_r28957538>:
>
> > +        if parsed.port:
> > +            netloc = hostname + u":" + six.text_type(parsed.port)
> > +        else:
> > +            netloc = hostname
> > +
> > +        # Note that building a URL in this fashion means it should be
> > +        # semantically indistinguishable from the original but is not
> > +        # guaranteed to be exactly the same.
> > +        uri = urllib_parse.urlunparse((
> > +            parsed.scheme,
> > +            netloc,
> > +            parsed.path,
> > +            parsed.params,
> > +            parsed.query,
> > +            parsed.fragment
> > +        ))
>
> urlparse requires // so if there's a valid URI without // it'll dump
> everything into the "path". So the failure mode will be that the data
> passes through without being parsed (and any IDNA won't be decoded).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1862/files#r28957538>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084